proposals/dht-js.md

```diff
@@ -0,0 +1,106 @@
+# Effective content and service discovery in JS
+
+Authors: @jacobheun
+
+Initial PR: TBD
+
+## Purpose &amp; impact
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives.
+-->
+
+**Current State**
+- JS projects rely on delegate and preload nodes to be able to interact with the live network
+- PL hosted delegate/preload nodes are often overloaded negatively impact performance of projects
```

Comment:
The requirement to preload content also makes JS projects very bandwidth heavy - every piece of data I add to a JS node is transferred to a preload node, this is very expensive in terms of transfer.

Reply:
Using DHT delegates suffers from similar problems, except it's sort of worse - we re-publish every hour so content survives the timed garbage collection, which means re-uploading every block, which scales poorly.

Reply:
I'm not sure I understand why delegated routing should be slow/expensive or what preload nodes are necessary in non-browser use cases.

DHT publishing frequency with delegated routing should match what you'd expect with non-delegated routing, no?

Reply:
Yes, the frequency should be the same, but as currently implemented with delegated routing, if a block has been garbage collected, you end up sending the data to the server each time for it to create the provider record - the whole DAG, not just the root block since the delegate has to be able to supply everything on behalf of your node.

> what preload nodes are necessary in non-browser use cases

They're not, at least, they shouldn't be - if the DHT implementation was complete and performant, which it isn't.  This proposal is to finish up the work we did durning the hack week and make it so.  Then we could turn off preload/delegate for everything that isn't a browser node.

Reply:
With proper delegated routing shouldn't I be making my own provider records that point at me instead of having ones that points at the preload nodes?

Reply:
> E.g. I ask the delegated router for the 20 closest peers then directly send them provider record puts.

Yes, but this is a limitation in the browser, as we will not be able to dial most of the network

Reply:
Sure, but my comment was about non-browser nodes not needing to store their data on preload nodes.

Browser nodes have a very rough time talking to the rest of the network at the moment, so they need to do a lot more delegating of work. Does anything in this proposal really help the browser node situation?